### DISCUSSION TITLE
ZIG: Auto-responder: ZbSend triggers Rule for ESW-0ZAA-EU Moes 4-button (ZigBee) Switch

### DISCUSSION THREAD

**Discussion by afgcons** [arendst/Tasmota#20337] [Q&A]
Each Switch has 4 buttons and each button decodes single click, double-click and long-press:
{"EndPoint":n}  n = 1..4 the button
{"LidlPower":n} n = 0..2 the key press 

I need to store at least one %value% in a var:

Rule3
 on zbreceived#0x2f5b#Endpoint do var2 %value% endon
 on zbreceived#0x2f5b#LidlPower do WebSend [usbnas.loc/cgi-bin/mocha.cgi] zbsw2x%var2%_%value% endon

The above version uses only one var and performs the WebSend on LidlPower instead of EndPoint. The previous version caused an old X-10 module to turn on/off every now and then and it turns out it wasn't noise on the power line. An Auto-responder ZbSend did it:

17:06:45.988 ZIG: Auto-responder: ZbSend {"Device":"0x2F5B","Cluster":"0x000A","Endpoint":1,"Response":"000A/0007":757098405}
17:06:45.993 MQT: tele/zb1/SENSOR = {"ZbReceived":{"0x2F5B":{"Device":"0x2F5B","Name":"zb2_4btnSw","Cluster":10,"Read":[7],"ReadNames":{"LocalTime":true},"Endpoint":1,"LinkQuality":48}}}
17:06:46.004 RUL: ZBRECEIVED#0X2F5B#ENDPOINT performs "var2 zbsw2x1_0"
17:06:46.007 MQT: stat/zb1/RESULT = {"Var2":"zbsw2x1_0"}
17:06:46.016 RUL: ZBRECEIVED#0X2F5B#LINKQUALITY performs "WebSend [usbnas.loc/cgi-bin/mocha.cgi] zbsw2x1_0"
17:06:46.163 MQT: stat/zb1/RESULT = {"WebSend":"Done"}

### My Questions:

1.  I have 3 Switches, using var1, var2, var3 respectively, is it possible  another, intervening event could alter the var's content?
2. Is there a trick to capture both %value% and events to avoid interference and false triggers?

I'm learning bit by bit from the Rules Cookbook and examples found here. If you have or know of an example of what I should do, please let me know. Thanks in advance !!

**Comment by barbudor** [comment]
If your zigbee gateway is running on ESP32, you'd better using berry rules that are a little more flexible such as allowing a AND of triggers

**Reply by sfromis** [reply]
As rules are very asynchronous, you cannot rely on any sequence to be uninterrupted, but you can follow what happens in the console.

For what you try to do, instead of things like the endpoint rule, you have `%zbdevice%`, `%zbgroup%`, `%zbcluster%` and `%zbendpoint%` directly available.
https://tasmota.github.io/docs/Rules/#rule-variables

Still, I concur with the recommendation using Berry, which is a **lot** more flexible. I'd even consider this a good reason to upgrade from a ESP8266-based coordinator to one based on ESP32.
